src/v/kafka/server/connection_context.cc

```diff
@@ -353,169 +356,137 @@ connection_context::reserve_request_units(api_key key, size_t size) {
+        // it is a one off event, but we need protection from it being
+        // nullptr
+        if (likely(_snc_quota_context)) {
+            _server.snc_quota_mgr().record_request_intake(
+              *_snc_quota_context, size);
+        }
+    }
+
+    auto sres = ss::make_lw_shared(std::move(sres_in));
+
+    auto remaining = size - request_header_size - hdr.client_id_buffer.size()
+                     - hdr.tags_size_bytes;
+    auto buf = co_await read_iobuf_exactly(conn->input(), remaining);
+    if (_server.abort_requested()) {
+        // _server._cntrl etc might not be alive
+        co_return;
+    }
+    auto self = shared_from_this();
+    auto rctx = request_context(
+      self, std::move(hdr), std::move(buf), sres->backpressure_delay);
+    /*
+     * we process requests in order since all subsequent requests
+     * are dependent on authentication having completed.
+     *
+     * the other important reason for disabling pipeling is because
+     * when a sasl handshake with version=0 is processed, the next
+     * data on the wire is _not_ another request: it is a
+     * size-prefixed authentication payload without a request
+     * envelope, and requires special handling.
+     *
+     * a well behaved client should implicitly provide a data stream
+     * that invokes this behavior in the server: that is, it won't
+     * send auth data (or any other requests) until handshake or the
+     * full auth-process completes, etc... but representing these
+     * nuances of the protocol _explicitly_ in the server makes its
+     * behavior easier to understand and avoids misbehaving clients
+     * creating server-side errors that will appear as a corrupted
+     * stream at best and at worst some odd behavior.
+     */
+
+    const auto correlation = rctx.header().correlation;
+    const sequence_id seq = _seq_idx;
+    _seq_idx = _seq_idx + sequence_id(1);
+    auto res = kafka::process_request(
+      std::move(rctx), _server.smp_group(), *sres);
+
+    /*
+     * first stage processed in a foreground.
+     *
+     * if the dispatch/first stage failed, then we need to
+     * need to consume the second stage since it might be
+     * an exceptional future.
+     */
+    auto dispatch = co_await ss::coroutine::as_future(
+      std::move(res.dispatched));
+    if (dispatch.failed()) {
+        try {
+            co_await std::move(res.response);
+        } catch (...) {
+            vlog(
+              klog.info,
+              "Discarding second stage failure {}",
+              std::current_exception());
+        }
+        vlog(
+          klog.info,
+          "Detected error dispatching request: {}",
+          dispatch.get_exception());
+        self->conn->shutdown_input();
+        sres->tracker->mark_errored();
+        co_return;
+    }
+
+    /**
+     * second stage processed in background.
+     */
+    ssx::background = ssx::spawn_with_gate_then(_server.conn_gate(), [&] {
```

Comment:
nit: isn't this `ssx::spawn_with_gate`?

Also capturing by reference here seems sketchy, the previous code captured by value (or moved values into the closure) and that feels like a safer thing todo, even though the lambda will always be executed without suspension points (for now).

Reply:
>nit: isn't this ssx::spawn_with_gate?

ahh, yeh. fixed.

> Also capturing by reference here seems sketchy, the previous code captured by value (or moved values into the closure) and that feels like a safer thing todo,  even though the lambda will always be executed without suspension points (for now).

as long as the lambda isn't a coroutine, then it's safe. i mean, you still have to know what you are capturing, from what I understand, it only captures what is actually used in the body of the lambda.

in any case, it's not really redpanda style, so i fixed it.

Reply:
> as long as the lambda isn't a coroutine, then it's safe.

That's not technically true right? It's more about the implementation of `ssx::spawn_with_gate_then`. This is scenario I was thinking about.

```
ss::future<> do_thingy(ss::lw_shared_ptr<flubber>);

template <typename Func>
ss::future<> do_helper_func(Func f) {
  co_await ss::now();
  // oops now foo has been freed.
  co_await f();
}


auto foo = ss::make_shared<flubber>();
ssx::background = do_helper_func([&] { return do_thingy(foo); })
```

You're assuming that the lambda is invoked before the return, which is probably always going to be true for ssx::spawn_with_gate, but anyways like you said we don't generally capture by reference.